recipe-client-addon/bootstrap.js

```diff
@@ -29,69 +38,146 @@ const DEFAULT_PREFS = {
+        log.error(`Error setting startup pref ${childPrefName}; pref type does not match.`);
+        continue;
+      }
+
+      switch (experimentPrefType) {
+        case Services.prefs.PREF_STRING:
+          defaultBranch.setCharPref(realPrefName, experimentBranch.getCharPref(childPrefName));
+          break;
+
+        case Services.prefs.PREF_INT:
+          defaultBranch.setIntPref(realPrefName, experimentBranch.getIntPref(childPrefName));
+          break;
+
+        case Services.prefs.PREF_BOOL:
+          defaultBranch.setBoolPref(realPrefName, experimentBranch.getBoolPref(childPrefName));
+          break;
+
+        case Services.prefs.PREF_INVALID:
+          // This should never happen.
+          log.error(`Error setting startup pref ${childPrefName}; pref type is invalid (${experimentPrefType}).`);
+          break;
+
+        default:
+          // This should never happen either.
+          log.error(`Error getting startup pref ${childPrefName}; unknown value type ${experimentPrefType}.`);
+      }
+    }
+  },
+
+  observe(subject, topic, data) {
+    if (topic === UI_AVAILABLE_NOTIFICATION) {
+      Services.obs.removeObserver(this, UI_AVAILABLE_NOTIFICATION);
+      ShieldRecipeClient.startup();
+    }
+  },
+
+  install() {
+    // Nothing to do during install
+  },
+
+  startup(data, reason) {
+    // Initialization that needs to happen before the first paint on startup.
+    this.initShieldPrefs(DEFAULT_PREFS);
+    this.initExperimentPrefs();
+
+    // If the app is starting up, wait until the UI is available before finishing
+    // init.
+    if (reason === REASON_APP_STARTUP) {
+      Services.obs.addObserver(this, UI_AVAILABLE_NOTIFICATION);
+    } else {
+      ShieldRecipeClient.startup();
+    }
+  },
+
+  async shutdown(data, reason) {
+    // Wait for async write operations during shutdown before unloading modules.
+    await ShieldRecipeClient.shutdown(reason);
+
+    // In case the observer didn't run, clean it up.
+    Services.obs.removeObserver(this, UI_AVAILABLE_NOTIFICATION);
+
+    // Unload add-on modules. We don't do this in ShieldRecipeClient so that
+    // modules are not unloaded accidentally during tests.
+    let modules = [
+      "lib/ActionSandboxManager.jsm",
+      "lib/Addons.jsm",
+      "lib/AddonStudies.jsm",
+      "lib/CleanupManager.jsm",
+      "lib/ClientEnvironment.jsm",
+      "lib/FilterExpressions.jsm",
+      "lib/EventEmitter.jsm",
+      "lib/Heartbeat.jsm",
+      "lib/LogManager.jsm",
+      "lib/NormandyApi.jsm",
+      "lib/NormandyDriver.jsm",
+      "lib/PreferenceExperiments.jsm",
+      "lib/RecipeRunner.jsm",
+      "lib/Sampling.jsm",
+      "lib/SandboxManager.jsm",
+      "lib/ShieldPreferences.jsm",
+      "lib/ShieldRecipeClient.jsm",
+      "lib/Storage.jsm",
+      "lib/Uptake.jsm",
+      "lib/Utils.jsm",
+    ].map(m => `resource://shield-recipe-client/${m}`);
+    modules = modules.concat([
```

Comment:
I know you're just moving this, but I think this would be easier to read and faster if written as:

```js
modules.push(
   "resource://shield-recipe-client-content/AboutPages.jsm",
   "resource://shield-recipe-client-vendor/mozjexl.js"
);
```